llvm-external-projects/iree-dialects/lib/Dialect/IREEPyDM/Transforms/Rtl/LowerToRtlPass.cpp

```diff
@@ -0,0 +1,170 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "../PassDetail.h"
+#include "iree-dialects/Dialect/IREE/IREEDialect.h"
+#include "iree-dialects/Dialect/IREE/IREEOps.h"
+#include "iree-dialects/Dialect/IREEPyDM/IR/Ops.h"
+#include "iree-dialects/Dialect/IREEPyDM/Transforms/Passes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/BuiltinDialect.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+using namespace mlir::iree_pydm;
+
+namespace pydm_d = mlir::iree_pydm;
+
+namespace {
+
+class RtlFunc {
+ protected:
+  static FunctionType makeRaisingSignature(Builder b, ArrayRef<Type> inputs,
+                                           Type output) {
+    return b.getType<FunctionType>(
+        inputs, TypeRange{b.getType<pydm_d::ExceptionResultType>(), output});
+  }
+};
+
+template <typename RtlFuncTy>
+Operation *importRtlFunc(SymbolTable &symbolTable) {
+  OpBuilder builder(symbolTable.getOp()->getContext());
+  auto name = builder.getStringAttr(RtlFuncTy::getRtlName());
+  auto *existing = symbolTable.lookup(name);
+  if (existing) return existing;
+
+  // Does not exist - create detached and insert.
+  FunctionType signature = RtlFuncTy::getRtlSignature(builder);
+  OperationState state(symbolTable.getOp()->getLoc(),
+                       pydm_d::FuncOp::getOperationName());
+  pydm_d::FuncOp::build(builder, state, name, signature);
```

Comment:
a trick with the symbol table is that inserting again is a no-op but you still get name uniqueing - so you can use a normal OpBuilder::create<> and then just symbolTable.insert() to get the naming - no need to create detached.

Reply:
Yeah, for those case, I actually need the naming to match up properly. I had it written the other way but it involved more reaching in and finding the right insertion point and this was simpler. Both are a bit obtuse.